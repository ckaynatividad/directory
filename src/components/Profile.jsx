import React from 'react';
import { useHistory } from 'react-router-dom';

export default function Profile(profile) {
  const history = useHistory();

  const handleSubmit = () => {
    history.push('/profile/edit');
  };
  return (
    <div>
      Profile <button onClick={handleSubmit}>edit</button>
      <p>Name: {profile.name}</p>
      <p>Email: {profile.email}</p>
      <p>Birthday: {profile.birthday}</p>
      <p>Bio: {profile.bio}</p>
    </div>
  );
}
