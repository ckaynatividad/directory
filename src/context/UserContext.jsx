import { createContext, useContext, useMemo, useState } from 'react';
import { getUser } from '../services/users';

export const UserContext = createContext();

export function UserProvider({ children }) {
  const currentUser = getUser();
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
  );

  const value = useMemo(() => ({ user, setUser }), [user]);

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
}

export function useUser() {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('user provider needed');
  }

  return context;
}
