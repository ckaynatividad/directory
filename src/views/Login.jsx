import classNames from 'classnames';
import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import AuthForm from '../components/AuthForm';
import { signInUser, signUpUser } from '../services/users';

export default function Login({ setUser }) {
  const [type, setType] = useState('Login');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errorMsg, setErrorMsg] = useState('');
  const history = useHistory();

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      let resp;
      if (type === 'Login') {
        resp = await signInUser(email, password);
        console.log('logging in');
        history.replace('/profile');
        window.location.reload();
      } else {
        resp = await signUpUser(email, password);
        setUser(resp);
        history.push('/confirm-email');
      }
    } catch {
      setErrorMsg('error with login');
    }
  };
  return (
    <div>
      <h1>{type}</h1>
      <AuthForm
        email={email}
        setEmail={setEmail}
        password={password}
        setPassword={setPassword}
        errorMsg={errorMsg}
        handleSubmit={handleSubmit}
      />
      <button
        onClick={() => {
          setType('Sign Up');
        }}
        className={classNames({ active: type === 'Sign Up' })}
      >
        Sign Up
      </button>
    </div>
  );
}
