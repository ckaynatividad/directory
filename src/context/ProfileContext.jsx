import { createContext, useEffect, useMemo, useState } from 'react';
import { Redirect, useHistory } from 'react-router-dom';
import { useContext } from 'react/cjs/react.development';
import { getProfile } from '../services/profiles';
import { useUser } from './UserContext';

export const ProfileContext = createContext();

export function ProfileProvider({ children }) {
  const [loading, setLoading] = useState(true);
  const { user } = useUser();
  const [profile, setProfile] = useState({});
  console.log(loading);
  console.log('context', profile);
  useEffect(() => {
    setLoading(true);
    getProfile()
      .then((data) => setProfile(data))
      .finally(() => setLoading(false));
  }, [user?.id]);

  // const value = useMemo(
  //   () => ({ profile, setProfile, loading, setLoading }),
  //   [profile]
  // );
  const value = { profile, setProfile, loading, setLoading };

  return (
    <ProfileContext.Provider value={value}>{children}</ProfileContext.Provider>
  );
}

export const useProfile = () => {
  const context = useContext(ProfileContext);

  if (context === undefined) {
    throw new Error('profileprovider needed');
  }

  return context;
};
