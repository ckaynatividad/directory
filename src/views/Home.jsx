import React from 'react';
import { Link, Redirect } from 'react-router-dom';
import { useUser } from '../context/UserContext';

export default function Home() {
  const { user } = useUser();
  return (
    <div className="Home">
      {user.id && <Redirect to="/profile" />}

      {!user.id && (
        <>
          <h4>Huwwo</h4>
          <p>
            {' '}
            pwease <Link to="/login">sign in</Link>
          </p>
        </>
      )}
    </div>
  );
}
