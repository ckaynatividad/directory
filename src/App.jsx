import { BrowserRouter, Route, Switch } from 'react-router-dom';
import './App.css';
import Header from './components/Header';
import PrivateRoute from './components/PrivateRoute';
import Profile from './components/Profile';
import ProfileForm from './components/ProfileForm';
import { ProfileContext, ProfileProvider } from './context/ProfileContext';
import { UserProvider } from './context/UserContext';
import Confirm from './views/Confirm';
import Home from './views/Home';
import Login from './views/Login';
import ProfileCreate from './views/ProfileCreate';
import ProfileEdit from './views/ProfileEdit';
import ProfileView from './views/ProfileView';

export default function App() {
  return (
    <div className="App">
      <UserProvider>
        <ProfileProvider>
          <BrowserRouter>
            <Switch>
              <Route exact path="/">
                <Home />
              </Route>
              <Route exact path="/login">
                <Login />
              </Route>
              <Route exact path="/confirm-email">
                <Confirm />
              </Route>
              <PrivateRoute exact path="/profile">
                <Header />
                <ProfileView />
              </PrivateRoute>
              <PrivateRoute exact path="/profile/edit">
                <ProfileEdit />
              </PrivateRoute>
              <PrivateRoute exact path="/profile/create">
                <Header />
                <ProfileCreate />
              </PrivateRoute>
            </Switch>
          </BrowserRouter>
        </ProfileProvider>
      </UserProvider>
    </div>
  );
}
