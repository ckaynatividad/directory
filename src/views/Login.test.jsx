import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import App from '../App';
import Profile from '../components/Profile';
import { UserProvider } from '../context/UserContext';
import Home from './Home';
import Login from './Login';
jest.mock('../context/UserContext');
test('login render', () => {
  render(
    <MemoryRouter>
      <UserProvider>
        <Login />
      </UserProvider>
    </MemoryRouter>
  );
  const login = screen.getByText(/Login/i);
  expect(login).toBeInTheDocument();
});

test('sign up button works', () => {
  render(
    <MemoryRouter>
      <UserProvider>
        <Login />
      </UserProvider>
    </MemoryRouter>
  );
  const signup = screen.getByRole('button', { name: /Sign Up/i });
  userEvent.click(signup);

  expect(screen.getByRole('heading', { name: /Sign Up/i })).toBeInTheDocument();
});

test('login works', () => {
  const { container } = render(
    <UserProvider mockUser={{ id: 1, email: 'natividad.ckay@gmail.com' }}>
      <MemoryRouter initialEntries={['/login']}>
        <Login />
      </MemoryRouter>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});
