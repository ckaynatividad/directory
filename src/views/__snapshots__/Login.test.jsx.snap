// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login works 1`] = `
<div>
  <div>
    <h1>
      Login
    </h1>
    <div>
      <form>
        <label>
          Email:
          <input
            type="email"
            value=""
          />
        </label>
        <label>
          Password:
          <input
            type="password"
            value=""
          />
        </label>
        <p />
        <input
          type="submit"
        />
      </form>
    </div>
    <button
      class=""
    >
      Sign Up
    </button>
  </div>
</div>
`;
